### ISSUE TITLE
Consider multipart operators

### ISSUE THREAD

**Issue by masak** [masak/007#18]
The prototypical case of this is `?? !!`, so let's start there.

What if we could define this operator in 007 like this:

    macro infix:<?? !!>(cond, then_val, else_val) {
        # elided implementation; it's interesting but not relevant here
    }

The main point of which is that instead of the usual two, an infix of this kind would have three parameters.

(Only reason it has to be a macro is to get thunkish semantics on `then_val` and `else_val`. We might still end up with a neater solution for that which doesn't need to involve outright macros. We'll see.)

And this would generalize not just to two-part infix operators, but to N-part infix operators. An N-part infix operator would require N+1 parameters. (And in fact, we might as well have a compile-time check for this. In fact, that's entirely orthogonal to the rest of this issue and probably a good idea anyway.)

From what I can see, allowing expressions *inside* of an infix operator like that has the following consequences:

* **The other shoe might never drop.** If the operator is defined like `infix:<?? !!>` and the code looks like `2 ?? 3;`, then the compiler needs to flag up an error that the `?? !!` operator wasn't completed. (Prior art in Perl 6: `Found ?? but no !!`. Although this error is hard-coded for the `?? !!` operator, since in Perl 6 there is no general mechanism as described here.)

* **Lower-precedence operators must be forbidden in the in-between expressions.** It's OK for a little expression tree to form between the `??` and the `!!`, but the operators used in that expression tree *must* be tighter than `?? !!` itself, for the simple reason that if they were looser, then `?? !!` couldn't be reduced before them.

* **Associativity still exists, but is prejudiced towards the first and last expressions.** "Left-associative" means that the operator makes subtrees in the first subexpression, and so `A ?? B !! C ?? D !! E` would mean `(A ?? B !! C) ?? D !! E`. "Right-associative" means that the operator makes subtrees in the last subexpression, and so `A ?? B !! C ?? D !! E` would mean `A ?? B !! (C ?? D !! E)`. (As indeed it does in Perl 6, since `?? !!` associates to the right.) There is no mechanism to make the operator associate through the middle subexpression, for the simple reason that that doesn't solve the problem of evaluation order.

* **This setup could enable an interesting and natural kind of multi dispatch.**

Let me be more specific about the last point. Say we wanted to implement a `??` operator, with the semantics that `A ?? B` worked like `A ?? B !! None`. The *conservative* thing to do would be to say that we can't have `??` and `?? !!` in the same environment, since they have the same first part. But the *interesting* thing to do would be to allow both, and simply keep both options open as we parse.

*(Note: it feels to me like this would work, and not bend the expression parser beyond the breaking point. Some of the points above about error messages get seriously muddled up (just like with multi subs vs only subs), but that looks merely hard to me, not insurmountable. In the worst case I'm wrong, and the resulting expression parser would have to do exponential backtracking or something. In which case we should revert to being conservative rather than interesting.)*

The interplay between this idea and `is parsed` is not clear to me. This may or may not be a problem, depending how much you like `is parsed` (or have an alternative up your sleeve). The closest I can come to an actual suggestion here is that `is parsed` needs to be free-form enough while also respecting the "fixed" bits of the operator. Or maybe that's not true at all and `is parsed` is in total control and the parser ignores everything except the first part of the `<?? !!>` specification.

Multipart prefixes and postfixes? I don't see why not, but I also suspect they would be less in demand than their infix counterparts. I haven't been able to come up with a use case for either. *(Edit: actually, postcircumfixes are an excellent use case, and even have the multi-part syntax in Perl 6, although the number of parts is constrained to 2.)*

**Comment by vendethiel** [comment]
Somewhat related but more complicated: [Parsing mixfix operators](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.157.7899)

**Comment by masak** [comment]
Interesting paper. Definitely touches on similar ideas. I'm not ready to make the precedence relation a DAG just yet &mdash; total order works fine for my purposes. ☺

But one part reminds me of a thing I didn't mention in the OP: if we decide to make *terms* definable as subs in the language too, we could get away with having parentheses being user-definable as the multipart `macro term:<( )>`. The macro body would simply extract the expression in-between the two fixed `(` and `)` strings, and return its Qtree.

**Comment by masak** [comment]
Not only `macro term:<( )>` is an attractive use case for this feature, but also `sub postfix:<[ ]>` and `macro postfix:<( )>`. These are currently special-cased in the parser, and it would be a big win to be able to un-special-case them.

Among the two, `sub postfix:<[ ]>` is the easier one. `macro postfix:<( )>` would require something like `is parsed` &mdash; but again, would be an excellent test case for it &mdash; since inside those invocation parentheses is the only place in the language where comma (`,`) is allowed. Furthermore, it then falls on `macro postfix:<( )>` to decide whether the call in question is a sub invocation or a macro invocation.

In other words, whatever this feature becomes, it should serve the in-language definition of array indexing and call syntax. If it can't do that, then this issue should be closed in disgrace.

**Comment by masak** [comment]
I think we can close this one for now; I see it being handled in other ways, mostly with a goodenuf `is parsed` mechanism. We *do* want to handle `postfix:<[ ]>` and `postfix:<( )>` natively. But doing it with multipart operators is probably overkill.
